### ISSUE TITLE
FR: support using self-signed IP-only certificate in custom DERP server, fingerprints for self-signed cert validation

### ISSUE THREAD

**Issue by networkhermit** [tailscale/tailscale#11776]
### What are you trying to do?

Not everyone has a domain name plus dns hosting. It's more accessible for users if we support using self-signed ip-only certificate in custom DERP server (just like k8s 6443 tls).

This issue had been brought up by #3647 and partially in #4082.

This feature is kind of implemented by a fork, [ip_derper](https://github.com/yangchuansheng/ip_derper) (which is regularly updated and just adding the patch for ip-based certificates):

https://github.com/yangchuansheng/ip_derper/blob/936de2789923d81b9eab538b50f124110dee6f9b/.github/workflows/main.yml#L21-L24

```yaml
      - name: modify
        run: |
          cd tailscale
          sed -i '/hi.ServerName != m.hostname/,+2d' cmd/derper/cert.go
```

The above modification is not ideal for users who want to use domain-based tls certificates. I think some more appropriate minor fix could be the following:
https://github.com/tailscale/tailscale/blob/226486eb9ae90a32c3bb6857b9dbc3dec605597c/cmd/derper/cert.go#L90-L93
```diff
 func (m *manualCertManager) getCertificate(hi *tls.ClientHelloInfo) (*tls.Certificate, error) {
-       if hi.ServerName != m.hostname {
+       if hi.ServerName != m.hostname && !(hi.ServerName == "" && net.ParseIP(m.hostname) != nil) {
                return nil, fmt.Errorf("cert mismatch with hostname: %q", hi.ServerName)
        }
```

So we could support well both domain-based certificates and the ip-based ones.

### What is the impact of not solving this?

Separate fork keeps patching this feature and some users fallback to that.

**Comment by bradfitz** [comment]
Sorry, but this isn't a use case we want to support.

You're welcome to modify the Tailscale client or derper code for your own use.

**Comment by networkhermit** [comment]
Could you elaborate a bit more? As this feature is regularly brought up by other users.

If the official reponse to this feature is "make your own fork", then related issues in the future could get linked to the explanation.

**Comment by bradfitz** [comment]
People running their own DERP servers is a constant source of support costs for us. We provide the source code for transparency reasons and in theory it saves us money if people run their own, and it's better for users if their DERP servers are lower latency and close to them. So we do, and it's open source, and we support custom DERP servers.

And yet: so many people run them in bizarre unsupported ways and come to us for help, which we spend lots of time helping with.

We do not want yet another possible configuration to support and maintain in the code base, especially for something so fringe: LetsEncrypt is free, and domain names are often nearly free. So I'm skeptical there's enough demand for this feature to justify the various costs.

**Comment by itsalongstory** [comment]
> LetsEncrypt is free, and domain names are often nearly free. So I'm skeptical there's enough demand for this feature to justify the various costs.

In some countries, domains without an ICP license are not permitted to provide web services. For individual users, obtaining an ICP license can be extremely cumbersome.

https://en.wikipedia.org/wiki/ICP_license

For these users, supporting deployment of self-signed certificates based on IP would be extremely convenient.

**Comment by ArcticLampyrid** [comment]
> so many people run them in bizarre unsupported ways and come to us for help, which we spend lots of time helping with.

I believe this only reflects the true demands of the community.


> So I'm skeptical there's enough demand for this feature to justify the various costs.

I'm sorry, but you are contradicting yourself by pointing out that too many people are seeking support on one hand, while stating that this is a rare need on the other.

**Comment by NathanSweet** [comment]
Domain names expose information that an IP does not.

**Comment by yafengabc** [comment]
The free certificate is only available for 3 months now, so it is very inconvenient to use certificates issued by institutions because they are only for encryption and not for other purposes. Why can't we use certificates made by ourselves?

**Comment by bradfitz** [comment]
Change in plans. We had a discussion and decided we'll do this.

We'll let you specify an optional TLS cert fingerprint in your DERP map on a node and use that, if present, to verify nodes instead of the normal WebPKI roots.  And then, yes, if you set your hostname to an IP address, we'll let you skip SNI too.

**Comment by aruiplex** [comment]
Hi, this is very good news for users under the GFW, I am looking forward to knowing more info on this FR.

**Comment by yafengabc** [comment]
This is a great change. After a simple test, I successfully deployed the derp server using a self-signed certificate.

**Comment by yafengabc** [comment]
> Hi, this is very good news for users under the GFW, I am looking forward to knowing more info on this FR.

https://www.cnblogs.com/yafengabc/p/18591222

**Comment by itsalongstory** [comment]
> > Hi, this is very good news for users under the GFW, I am looking forward to knowing more info on this FR.
> 
> 
> 
> https://www.cnblogs.com/yafengabc/p/18591222

Great!

**Comment by yqs112358** [comment]
> > > Hi, this is very good news for users under the GFW, I am looking forward to knowing more info on this FR.
> > 
> > 
> > https://www.cnblogs.com/yafengabc/p/18591222
> 
> Can `InsecureForTests` be removed from the config file?

No. Maybe we should wait for Tailscale to support specify fingerprints for self-signed certs to pass the TLS cert validation.

**Comment by GuoJiaxing1997** [comment]
> Change in plans. We had a discussion and decided we'll do this.
> 
> We'll let you specify an optional TLS cert fingerprint in your DERP map on a node and use that, if present, to verify nodes instead of the normal WebPKI roots. And then, yes, if you set your hostname to an IP address, we'll let you skip SNI too.

That is so good!!!!!

**Comment by bradfitz** [comment]
I sent https://github.com/tailscale/tailscale/pull/15208

**Comment by jsbyysheng** [comment]
I saw this feature has been released with the latest [1.82.0](https://tailscale.com/changelog):

> [DERP](https://tailscale.com/kb/1232/derp-servers) functionality within the client supports certificate pinning for self-signed IP address certificates for those unable to use [Let's Encrypt](https://letsencrypt.org/) or WebPKI certificates.

However, `tailscale status` told me:
```
Tailscale could not connect to the 'xxx' relay server. Your Internet connection might be down, or the server might be temporarily unavailable.
```
At the same time, `tailscale debug derp xxx` told me:
```
"Error upgrading connection to node \"x.x.x.x\" @ \"x.x.x.x:44443\" to TLS over IPv4: tls: failed to verify certificate: x509: certificate is not valid for any names, but wanted to match sha256-raw:xxxxxxxxxxxxxxxxxxxxxx"
```
Anyone have tried this new feature?

**Comment by AsterisMono** [comment]
@XuRainbow This is fixed in https://github.com/tailscale/tailscale/pull/16016. Not released yet

Also see https://github.com/tailscale/tailscale/issues/15579

**Comment by ghost** [comment]
Hi! Any news on releasing the self-cert version of Tailscale app?

**Comment by jsbyysheng** [comment]
> Hi! Any news on releasing the self-cert version of Tailscale app?

It runs properly on the latest version.

**Comment by ghost** [comment]
> > Hi! Any news on releasing the self-cert version of Tailscale app?
> 
> It runs properly on the latest version.

How? It still doesn't work, because I need to add my domain in the custom address field for Headscale. When I try to connect, I get error message that the connection was refused, and that the context got canceled.

I take it that this is because Tailscale app does not know that I run self-signed certificate, so it refuses my login. Have this function been tested properly?

I see an "auth key" section too, but this is something that needs to be generated by my console, and the preauth-key I generate from console doesn't get accepted by this section on the Tailscale app, it seems to be the wrong format.
I also think that it refuses it because still it doesn't trust my certificate.

**Comment by jsbyysheng** [comment]
> > > Hi! Any news on releasing the self-cert version of Tailscale app?
> > 
> > It runs properly on the latest version.
> 
> How? It still doesn't work, because I need to add my domain in the custom address field for Headscale. When I try to connect, I get error message that the connection was refused, and that the context got canceled.
> 
> I take it that this is because Tailscale app does not know that I run self-signed certificate, so it refuses my login. Have this function been tested properly?
> 
> I see an "auth key" section too, but this is something that needs to be generated by my console, and the preauth-key I generate from console doesn't get accepted by this section on the Tailscale app, it seems to be the wrong format.
> I also think that it refuses it because still it doesn't trust my certificate.

Why Headscale? You should use Tailscale Derper.

**Comment by bradfitz** [comment]
This issue is only about DERP servers, not the control plane.
